Updates for PDF (code not tested yet)

cippy opened the pull request:
Some changes in **makeHistogramsWmass.py** to deal with PDFs and alphaS variations (PDF details read from **wmassUtilities/theory.py**, so that we can be consistent with any other script).
Also, separating process and systematic name in the histogram name with a double underscore (just like mcPlots.py does), so to allow us to manipulate them more easily (and allow for nuisances with single underscores in their name).

Minor changes in some source files with functions to deal with projections, mainly in the choice of bins to project.
For instance, if we need a single bin we should pass binStart = binEnd, but this convention was not used consistently in all functions.

Not yet merging this PR, I might add new things today.

kdlong commented on WMass/python/plotter/ccFiles/systHistHelpers.cc:
```diff
@@ -15,7 +15,14 @@ T mirrorHist(T& nominal, T& alternate, T& mirrored) {
         double yM = (yA > 0 && y0 > 0) ? y0*y0/yA : 0;
         //std::cout << "Here yM is " << yM << " y0 is " << y0 << " ratio is " << y0/yA << std::endl;
         mirrored.SetBinContent(b, yM);
-        mirrored.SetBinError(b, alternate.GetBinError(b)*y0/yA);
```
I'm not sure I agree with this comment and it doesn't add much in any case...

cippy replied:
But is alternate.GetBinError(b) set anywhere? I think that if we keep "alternate.GetBinError(b)*y0/yA" we should actually use nominal.GetBinError(b)*y0/yA. Or, if alternative and nominal have the same uncertainties because of cloning the alternative from nominal, then so should mirror.

kdlong replied:
This assumes that the alternate hist was built from a proper estimation, if will have proper unc if it's produced by weights for example. The only reason I can see that you'd want the uncertainty of the alternate hist is to see if there is a high stat uncertainty somehow, and for that the alternate unc is more useful.

But fine, you have it for one or the other, I don't really care if you set it to nominal.GetBinError(b)

kdlong commented on WMass/python/plotter/makeHistogramsWMass.py:
```diff
@@ -166,37 +165,58 @@ def writeAndRemove(hists):
-        elif "pdfCT18" in syst:
-            writeAndRemove(util.makeVariationHistsForCharge(h3D, f"x_{proc}_"+"pdf{label}CT18", util.pairGroups(58)))
+            writeAndRemove(util.makeVariationHistsForCharge(h3D, f"x_{proc}__QCDscale{ptbin}{{label}}", qcdScaleIndices()))
+        # check if name contains "pdf" followed by one or more capital letters
```
I personally dislike comments that just explain what code is doing, because if you read the code it's obvious, and if someone changes code later and forgets the comment, it's misleading

cippy replied:
acknowledged

kdlong commented on WMass/python/plotter/mcAnalysis.py:
```diff
@@ -437,61 +437,65 @@ def getPlots(self,plotspecs,cut,process=None,nodata=False,makeSummary=False,clos
-            #return retyields
-            self.prettyPrint(retyields)
-        # done with yields
+        #
```
Do you plan to restore it in the future? Since it lives forever in git anyway, I prefer to delete vs. comment out

cippy replied:
Yes, it might just be removed, I don't plan to add back this feature, but first I wanted to make sure nothing else breaks after commenting.

kdlong replied:
What does that mean? What would break? It runs, doesn't it?

kdlong commented on WMass/python/plotter/wmassUtilities/theory.py:
```diff
@@ -20,19 +24,21 @@
         "weight" : "LHEPdfWeightAltSet18",
         "alphas" : ["LHEPdfWeightAltSet18[59]", "LHEPdfWeightAltSet18[60]"],
-        "alphaRange" : "002",
+        "alphaRange" : "001",
```
We should also change the value given in systsFit.txt to scale up the 0.01 value

cippy replied:
Right, I could update that one before merging, unless it clashes with what you are currently testing

kdlong replied:
Better to update now I think, so it's consistent. If there are conflicts with my stuff they'll be trivial.

kdlong commented on WMass/python/plotter/wmassUtilities/theory.py:
```diff
@@ -20,19 +24,21 @@
     "ct18" : {
         "name" : "CT18",
         "entries" : 59,
+        "nhessian" : 58,
```
If it's always entries-1, is this really needed?

cippy commented:
I removed some comments, and for the pdf dictionary (see **wmassUtilities/theory.py**) and its usage I think we can use the following convention.
Alpha is always dealt with separately wrt actual pdf hessians (we are defining alpha explicitly in the dictionary and putting it in a separate histogram), so in the dictionary I removed "entries" and left "nhessian". When building pdf histograms we will always consider the branch up to nhessian+1 elements (the first is the nominal, assuming it is always there).

Just one **caveat** for nnpdf31: 
when filling the histogram with all variations the whole branch is currently used, as always, but it also has the 2 alpha variations. These will just go in the overflow bin. I am not sure whether we may prefer to truncate the branch as we do for CT18, it should not change performance too much if we evaluate 2 additional elements that we eventually do not use out of a total of 103.
